File: libs/eave-stdlib-ts/src/core-api/operations/documents.ts
import { sharedConfig } from '../../config.js';
import { EaveApp } from '../../eave-origins.js';
import { CtxArg, RequestArgsOrigin, RequestArgsTeamId, makeRequest } from '../../requests.js';
import { DocumentInput, DocumentSearchResult } from '../models/documents.js';
import { DocumentReference, Subscription } from '../models/subscriptions.js';
import { Team } from '../models/team.js';
import { CoreApiEndpointConfiguration } from './shared.js';

export type UpsertDocumentRequestBody = {
  document: DocumentInput;
  subscriptions: Array<Subscription>;
}
export type UpsertDocumentResponseBody = {
  team: Team;
  subscriptions: Array<Subscription>;
  document_reference: DocumentReference;
}

export class UpsertDocumentOperation {
  static config = new CoreApiEndpointConfiguration({ path: "/documents/upsert" })

  static async perform(args: RequestArgsTeamId & {input: UpsertDocumentRequestBody}): Promise<UpsertDocumentResponseBody> {
    const resp = await makeRequest({
      url: this.config.url,
      ...args,
    });
    const responseData = <UpsertDocumentResponseBody>(await resp.json());
    return responseData;
  }
}

export type SearchDocumentsRequestBody = {
  query: string;
}

export type SearchDocumentsResponseBody = {
  team: Team;
  documents: DocumentSearchResult[];
}

export class SearchDocumentsOperation {
  static config = new CoreApiEndpointConfiguration({ path: "/documents/search" })

  static async perform(args: RequestArgsTeamId & {input: SearchDocumentsRequestBody}): Promise<SearchDocumentsResponseBody> {
    const resp = await makeRequest({
      url: this.config.url,
      ...args,
    });
    const responseData = <SearchDocumentsResponseBody>(await resp.json());
    return responseData;
  }
}

File: libs/eave-stdlib-py/src/eave/stdlib/core_api/operations/documents.py
from typing import Unpack
import uuid

from eave.stdlib.core_api.models.documents import DocumentSearchResult
from eave.stdlib.core_api.models.documents import DocumentInput
from . import BaseRequestBody, BaseResponseBody, CoreApiEndpoint, CoreApiEndpointConfiguration

from ..models.subscriptions import DocumentReference, DocumentReferenceInput, Subscription
from ..models.subscriptions import SubscriptionInput

from ..models import team
from ... import requests


class UpsertDocument(CoreApiEndpoint):
    config = CoreApiEndpointConfiguration(
        path="/documents/upsert",
        auth_required=False,
    )

    class RequestBody(BaseRequestBody):
        document: DocumentInput
        subscriptions: list[SubscriptionInput]

    class ResponseBody(BaseResponseBody):
        team: team.Team
        subscriptions: list[Subscription]
        document_reference: DocumentReference

    @classmethod
    async def perform(
        cls,
        input: RequestBody,
        team_id: uuid.UUID,
        **kwargs: Unpack[requests.CommonRequestArgs],
    ) -> ResponseBody:
        response = await requests.make_request(
            config=cls.config,
            input=input,
            team_id=team_id,
            **kwargs,
        )

        response_json = await response.json()
        return cls.ResponseBody(**response_json, _raw_response=response)


class SearchDocuments(CoreApiEndpoint):
    config = CoreApiEndpointConfiguration(
        path="/documents/search",
        auth_required=False,
    )

    class RequestBody(BaseRequestBody):
        query: str

    class ResponseBody(BaseResponseBody):
        team: team.Team
        documents: list[DocumentSearchResult]

    @classmethod
    async def perform(
        cls, input: RequestBody, team_id: uuid.UUID, **kwargs: Unpack[requests.CommonRequestArgs]
    ) -> ResponseBody:
        response = await requests.make_request(
            config=cls.config,
            input=input,
            team_id=team_id,
            **kwargs,
        )

        response_json = await response.json()
        return cls.ResponseBody(**response_json, _raw_response=response)


class DeleteDocument(CoreApiEndpoint):
    config = CoreApiEndpointConfiguration(
        path="/documents/delete",
        auth_required=False,
    )

    class RequestBody(BaseRequestBody):
        document_reference: DocumentReferenceInput

    class ResponseBody(BaseResponseBody):
        pass

    @classmethod
    async def perform(
        cls,
        input: RequestBody,
        team_id: uuid.UUID,
        **kwargs: Unpack[requests.CommonRequestArgs],
    ) -> ResponseBody:
        response = await requests.make_request(
            config=cls.config,
            input=input,
            team_id=team_id,
            **kwargs,
        )

        return cls.ResponseBody(_raw_response=response)
